### PR TITLE
sonic-server: init at 1.4.0

### PR DESCRIPTION
**Pull request by pleshevskiy**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/search/sonic-server/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    rev = "f5302f5c424256648ba0be32b3c5909d846821fe";
+    sha256 = "sha256-WebEluXijgJckZQOka2BDPYn7PqzPTsIcV2T380fxW8=";
+  };
+
+  cargoSha256 = "sha256-ObhKGjaIma6fUVUT3xadpy/GPYlnm0nKmRVxFmoePyQ=";
+
+  buildNoDefaultFeatures = true;
+  buildFeatures =
+    (lib.optional withJemallocAllocator "allocator-jemalloc")
+    ++ (lib.optional withChineseTokenizer "tokenizer-chinese");
```
Comment (SuperSandro2000) [inline pkgs/servers/search/sonic-server/default.nix:27]:
```suggestion
  buildFeatures = lib.optional withJemallocAllocator "allocator-jemalloc"
    ++ lib.optional withChineseTokenizer "tokenizer-chinese";
```

File: pkgs/servers/search/sonic-server/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{ lib
+, rustPlatform
+, llvmPackages
+, clang
+, fetchFromGitHub
+, withJemallocAllocator ? true
+, withChineseTokenizer ? true
+, ...
```
Comment (SuperSandro2000) [inline pkgs/servers/search/sonic-server/default.nix:8]:
```suggestion
```
Why would you want to turn them off? They don't seem to increase closure size.

File: pkgs/servers/search/sonic-server/default.nix [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    install -Dm444 -t $out/etc/sonic config.cfg
+    substitute \
+      ./examples/config/systemd.service $out/usr/lib/systemd/system/sonic-server.service \
+      --replace /bin/sonic $out/bin/sonic \
+      --replace /etc/sonic.cfg $out/etc/sonic/config.cfg
+  '';
+
+  meta = with lib; {
+    description = "Fast, lightweight and schema-less search backend.";
```
Comment (06kellyjac) [inline pkgs/servers/search/sonic-server/default.nix:55]:
```suggestion
    description = "Fast, lightweight and schema-less search backend";
```

contributing requires descriptions not end in punctuation

File: pkgs/servers/search/sonic-server/default.nix (hunk 2 above)
Comment (pleshevskiy) [reply pkgs/servers/search/sonic-server/default.nix:8]: this is a more convenient and intuitive way of configuring the package when we support features through parameters for `pkgs.sonic-server.override {}` than through `pkgs.sonic-server.overrideAttrs (oldAttrs: { })`

File: pkgs/servers/search/sonic-server/default.nix (hunk 1 above)
Comment (pleshevskiy) [reply pkgs/servers/search/sonic-server/default.nix]: Thank you! didn't know about that

File: pkgs/servers/search/sonic-server/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/servers/search/sonic-server/default.nix:8]: Yes, I know but why do we want them. Wouldn't it be even better if we enabled those features all the time?

Comment (pleshevskiy) [reply pkgs/servers/search/sonic-server/default.nix:8]: I haven't used these features on my server for three years. Maybe someone else doesn't need them either. But they are default features in sonic-server. So I turned them on by default to be expected.

I can remove them as you suggested and I will use overrideAttrs. but I don't understand where the boundary is in the parameters. Why in some cases we can do that and in some cases we can't.

Comment (SuperSandro2000) [reply pkgs/servers/search/sonic-server/default.nix:8]: > I haven't used these features on my server for three years.

but what happens if those features are unused? If they just sit there and do nothing and have no significant impact on the closure size than having those options here is pointless.

>  but I don't understand where the boundary is in the parameters. Why in some cases we can do that and in some cases we can't.

Ideally we want as little options as possible. Especially options which do not require extra dependencies and have little to no impact on the final output size are not valuable to add unless they disable features which have significant security impact.

Comment (pleshevskiy) [reply pkgs/servers/search/sonic-server/default.nix]: I've removed these parameters as you suggested

**Pull request merged**